components/raftstore/src/store/util.rs

```diff
@@ -905,6 +994,60 @@ impl RegionReadProgress {
         }
     }
 
+    // Consume the provided `LeaderInfo` to update `safe_ts` and return whether the provided
+    // `LeaderInfo` is same as ours
+    pub fn consume_leader_info(&self, mut leader_info: LeaderInfo) -> bool {
+        let mut core = self.core.lock().unwrap();
+        if leader_info.has_read_state() {
+            // It is okay to update `safe_ts` without checking the `LeaderInfo`, the `read_state`
+            // is guaranteed to be valid when it is published by the leader
+            let rs = leader_info.take_read_state();
+            let (apply_index, ts) = (rs.get_applied_index(), rs.get_safe_ts());
+            if apply_index != 0 && ts != 0 && !core.discard {
+                if let Some(ts) = core.update_safe_ts(apply_index, ts) {
+                    if !core.pause {
+                        self.safe_ts.store(ts, AtomicOrdering::Release);
+                    }
+                }
+            }
+        }
+        // whether the provided `LeaderInfo` is same as ours
+        core.leader_info.leader_term == leader_info.term
+            && core.leader_info.leader_id == leader_info.peer_id
+            && region_epoch_equal(&core.leader_info.epoch, leader_info.get_region_epoch())
```

Comment:
Why not check leaderInfo first?

Reply:
So that we can return the result directly instead of assigning the result to a temporary variable and return it at the end.

Reply:
Then why it's checked first in the original implementation?

Reply:
Just for robustness I think, the check is unnecessary since even it passes in this peer doesn't mean it can pass among majority peers. The `read_state` item should be promised valid when it is published by the leader instead of letting the receiver validated it with the receiver's limited information.

Reply:
/cc @5kbpers